spatial_cell_gpu.cpp

```diff
@@ -0,0 +1,1170 @@
+      list_with_replace_new_capacity = other.list_with_replace_new_capacity;
+      list_delete_capacity = other.list_delete_capacity;
+      list_to_replace_capacity = other.list_to_replace_capacity;
+      list_with_replace_old_capacity = other.list_with_replace_old_capacity;
+
+      // Member variables
+      ioLocalCellId = other.ioLocalCellId;
+      sysBoundaryFlag = other.sysBoundaryFlag;
+      sysBoundaryLayer = other.sysBoundaryLayer;
+      sysBoundaryLayerNew = other.sysBoundaryLayerNew;
+      initialized = other.initialized;
+      mpiTransferEnabled = other.mpiTransferEnabled;
+      for (unsigned int i=0; i<bvolderivatives::N_BVOL_DERIVATIVES; ++i) {
+         derivativesBVOL[i] = other.derivativesBVOL[i];
+      }
+      for (unsigned int i=0; i<CellParams::N_SPATIAL_CELL_PARAMS; ++i) {
+         parameters[i] = other.parameters[i];
+      }
+      for (unsigned int i=0; i<WID3; ++i) {
+         null_block_data[i] = 0.0;
+      }
+      for (unsigned int i=0; i<MAX_NEIGHBORS_PER_DIM; ++i) {
+         neighbor_block_data[i] = 0;
+         neighbor_number_of_blocks[i] = 0;
+      }
+      face_neighbor_ranks.clear();
+      // for (unsigned int i=0; i<MAX_NEIGHBORS_PER_DIM; ++i) {
+      //    neighbor_block_data[i] = other.neighbor_block_data[i];
+      //    neighbor_number_of_blocks[i] = other.neighbor_number_of_blocks[i];
+      // }
+      // if (other.face_neighbor_ranks.size()>0) {
+      //    face_neighbor_ranks = std::map<int,std::set<int>>(other.face_neighbor_ranks);
+      // }
+      if (other.populations.size()>0) {
+         populations = std::vector<spatial_cell::Population>(other.populations);
+      }
+   }
+
+   const SpatialCell& SpatialCell::operator=(const SpatialCell& other) {
+      // Used for refining spatial cells
+      velocity_block_with_content_list_capacity = other.velocity_block_with_content_list_capacity;
+      velocity_block_with_content_list->clear();
+      velocity_block_with_content_list->reserve(velocity_block_with_content_list_capacity);
+      velocity_block_with_content_list_size = 0;
+      dev_velocity_block_with_content_list = velocity_block_with_content_list->upload<false>();
+
+      gpuStream_t stream = gpu_getStream();
+
+      if (vbwcl_sizePower < other.vbwcl_sizePower) {
+         vbwcl_sizePower = other.vbwcl_sizePower;
+         ::delete velocity_block_with_content_map;
+         void *buf1 = malloc(sizeof(Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>));
+         velocity_block_with_content_map = ::new (buf1)Hashinator::Hashmap<vmesh::GlobalID,vmesh::LocalID>(vbwcl_sizePower);
+         dev_velocity_block_with_content_map = velocity_block_with_content_map->upload<true>(stream);
+      } else {
+         velocity_block_with_content_map->clear<false>(Hashinator::targets::device,stream,std::pow(2,vbwcl_sizePower));
+      }
+      if (vbwncl_sizePower < other.vbwncl_sizePower) {
+         vbwncl_sizePower = other.vbwncl_sizePower;
```

Comment:
vbwcl and vbwncl is... a bit cryptic. But I guess I figured what it stands for, and is maybe OK this deep in the code. Also I'm 166 files into this PR and my cortex is starting to delaminate.

Reply:
Admittedly obscure variable naming, but renaming now would be a bit dangerous.